### ISSUE TITLE
Provide a given palette

### ISSUE THREAD

**Issue by SirVer** [Ivordir/quantette#8]
First, thank you for your work! This is a very useful contribution to the Rust ecosystem and is very well crafted!

I am trying to use this crate for the following: I have an E-Ink display that can display white, black and yellow, but no other colors. I would like to get a given image broken down to display as well as possible on this particular display. So I cannot use the kmeans clustering to get the best possible palette to choose from, instead I want to provide a palette. I still would like to use the Floyd Steinberg to get quality improvements through dithering. I was hoping for something like this:

~~~rust
let custom_palette = [
    Srgb::new(0.0, 0.0, 0.0), // Black
    Srgb::new(1.0, 1.0, 1.0), // White
    Srgb::new(1.0, 1.0, 0.0), // Yellow, the third color of my display.
];

let mut pipeline = quantette::ImagePipeline::try_from(image)?;
let quantized_image = pipeline
    .use_palette(custom_palette)  // <- This function currently does not exist.
    .colorspace(quantette::ColorSpace::Oklab)
    .dither(true)
    .quantized_rgbimage();
~~~

Maybe this is already possible and I just do not understand how? Maybe this is also difficult or impossible? Otherwise, this would be a very useful addition indeed.

**Comment by Ivordir** [comment]
Thanks for the issue, glad you're finding this useful!

Providing a custom palette is currently not supported/possible. But this shouldn't be too hard to add -- I'll get around to it sometime soon-ish.

**Comment by lublak** [comment]
This would be awesome! 👍

**Comment by IanManske** [comment]
Most quantization use cases I had in mind like PNG indexed color, GIF, and pixel art/retro graphics are limited to 256 colors. More than 256 colors requires two bytes per pixel (which eats into the space efficiency of indexed color) or bit packing (which is a little more complicated).

From [Wikipedia](https://en.wikipedia.org/wiki/Indexed_color):
> Indexed color images with palette sizes beyond 256 entries are rare. The practical limit is around 12-bit per pixel, 4,096 different indices. To use indexed 16 bpp or more does not provide the benefits of the indexed color images' nature, due to the color palette size in bytes being greater than the raw image data itself

So, I assumed a max of 256 colors in the design of `quantette` which allowed a bunch of optimizations. Supporting more than 256 colors would require:
1. The API **everywhere** to either be generic over the index type or duplicated for large palettes/indices.
2. Me to revisit a bunch of places and add optimizations for large palettes. E.g., the `NearestNeighborColorMap` uses a brute force search, but for large palettes it would probably need to be switched to a kd-tree implementation.

TLDR: I'm not sure it's worth the effort.

If you have a reference palette larger than 256, I can suggest you either:
1. Use the reference palette as input to `Pipeline` and generate a new palette. (This can modify all the palette colors which might not be what you want.)
2. If preserving the original palette colors is important, you can instead calculate the distances between all pairs of colors in your palette. Then, you can take the pair of points closest to one another and either merge them or delete one the colors. Repeat until you have 256 colors. I'm sure there's other smarter methods that you can do or come up with.

Hope this helps/explains things!

**Comment by SirVer** [comment]
Only got around to trying this today. Works like a charm! Thank you very much for implementing!
